packages/fetchai/connections/http/connection.py

```diff
@@ -0,0 +1,384 @@
+
+        """
+        self._client_id = uuid4().hex
+        protocol_id = PublicId.from_str("fetchai/http:0.1.0")
+        uri = URI(f"http://{self.host}:{self.port}{url}")
+        context = EnvelopeContext(connection_id=self.connection_id, uri=uri)
+        # Prepare the Envelope's message body and encode it into bytes.
+        msg = {
+            "performative": http_method,
+            "path": url,
+            "params": param,
+            "payload": body,
+        }
+        msg_bytes = json.dumps(msg).encode()
+        # Prepare the Envelope itself using the provided variables.
+        envelope = Envelope(
+            to=self.address,
+            sender=self._client_id,
+            protocol_id=protocol_id,
+            context=context,
+            message=msg_bytes,
+        )
+        # Send the Envelope to the Agent's InBox.
+        assert self.in_queue is not None, "Input queue not initialized."
+        assert self.loop is not None, "Loop not initialized."
+        asyncio.run_coroutine_threadsafe(self.in_queue.put(envelope), self.loop)
+
+    def _send_response(self):
+        """Pass the response from out-bounded Envelope back to the front-end client.
+
+        Currently, the response will be written back in the cmd terminal.
+        """
+        # self.wfile.write(self.path.encode())
+        pass
+
+    def connect(self):
+        """
+        Connect.
+
+        Upon HTTP Channel connection, kickstart the HTTP Server.
+        """
+        with self.lock:
+            if self.stopped:
+                self.stopped = False
+                # self.thread.start()
+                # Initializing the internal HTTP server
+                try:
+                    # self.httpd = HTTPServer((self.host, self.port), HTTPHandler(self))
+                    # handler = SimpleHTTPServer.SimpleHTTPRequestHandler
+                    # handler = HTTPHandler.set_channel(self)
+                    # handler.set_channel(self)
+                    self.httpd = HTTPServer(
+                        (self.host, self.port), HTTPHandlerFactory(self)
+                    )
+                    # self.httpd = HTTPServer((self.host, self.port), handler)
+                    logger.info(f"HTTP Server has connected to port {self.port}.")
+                    self.httpd.serve_forever()
+                except OSError:
+                    logger.error(
+                        f"{self.host}:{self.port} is already in use, please try another Socket."
+                    )
+
+    def disconnect(self) -> None:
+        """
+        Disconnect.
+
+        Shut-off the HTTP Server and then HTTP Channel.
+        """
+        with self.lock:
+            if not self.stopped:
+                self.httpd.shutdown()
+                logger.info(f"HTTP Server has shutdown on port {self.port}.")
+                self.stopped = True
+                # self.thread.join()
+
+
+def HTTPHandlerFactory(channel: HTTPChannel):
```

Comment:
Is the factory pattern needed here? Also, PascalCase is never used for functions afaik.

Reply:
Yes it is needed, otherwise, I would not be able to pass in the channel arg into the handler when initializing HTTPServer: `self.httpd = HTTPServer((self.host, self.port), HTTPHandlerFactory(self))`. I tried without factory but failed because HTTPHandler also need to inherit from BaseHTTPRequestHandler.

I am following the  **MakeHandlerClassFromArgv** example here, including the naming convention:

https://stackoverflow.com/questions/21631799/how-can-i-pass-parameters-to-a-requesthandler

Reply:
ok